File: Source/UIViewController+ChildController.m
//
//  UIViewController+ChildController.m
//  JCLightWeibo
//
//  Created by Jake on 18/04/2017.
//  Copyright © 2017 Jake Cai. All rights reserved.
//

#import "UIViewController+ChildController.h"

@implementation UIViewController (ChildController)

-(void)addChildVC:(UIViewController *)viewController
{
    [self addChildViewController:viewController frame:self.view.bounds];
}

- (void)addChildViewController:(UIViewController *)viewController
                        inView:(UIView *)view
                     withFrame:(CGRect)frame
{
    [self addChildViewController:viewController
                setSubViewAction:^(UIViewController *superViewController,
                                   UIViewController *childViewController) {
                    childViewController.view.frame = frame;
        
                    if (![view.subviews containsObject:childViewController.view]) {
                        [view addSubview:childViewController.view];
                    }
                }];
}

- (void)addChildViewController:(UIViewController *)viewController
                         frame:(CGRect)frame
{
    [self addChildViewController:viewController
                setSubViewAction:^(UIViewController *superViewController,
                                   UIViewController *childViewController) {
                    childViewController.view.frame = frame;
                    
                    if (![superViewController.view.subviews containsObject:viewController.view]) {
                        [superViewController.view addSubview:viewController.view];
                    }
                }];
}

- (void)addChildViewController:(UIViewController *)viewController
              setSubViewAction:(void(^)(UIViewController *superViewController,
                                        UIViewController *childViewController))setSubViewAction
{
    BOOL containsVC = [self.childViewControllers containsObject:viewController];
    
    if (!containsVC) {
        [self addChildViewController:viewController];
    }
    
    if (setSubViewAction) {
        setSubViewAction(self , viewController);
    }
    
    if (!containsVC) {
        [viewController didMoveToParentViewController:self];
    }
    
}

- (void)removeFromParentVC
{
    [self willMoveToParentViewController:nil];
    [self.view removeFromSuperview];
    [self removeFromParentViewController];
}

@end

File: Source-swift/UIViewController+ChildController.swift
//
//  UIViewController+ChildController.swift
//  JCPhotoBrowser-Swift
//
//  Created by Jake on 31/05/2017.
//  Copyright © 2017 Jake. All rights reserved.
//
import UIKit

extension UIViewController {
    func addChildVC(_ viewController:UIViewController) {
        self.addChildViewController(viewController,frame: self.view.bounds)
    }
    
    func addChildViewController(_ viewController:UIViewController,inView:UIView,withFrame:CGRect) {
        self.addChildViewController(viewController) { (superViewController,childViewController) in
            childViewController.view.frame = withFrame;
            
            if inView.subviews.contains(viewController.view) == false {
                inView.addSubview(viewController.view)
            }
        }
    }
    
    func addChildViewController(_ viewController:UIViewController,frame:CGRect) {
        self.addChildViewController(viewController) { (superViewController,childViewController) in
            childViewController.view.frame = frame;
            
            if superViewController.view.subviews.contains(viewController.view) == false {
                superViewController.view.addSubview(viewController.view)
            }
        }
    }
    
    func addChildViewController(_ viewController:UIViewController,
                                   setSubViewAction:((_ superViewController:UIViewController,_ childViewController:UIViewController) -> Void)?) {
        
        let containsVC = self.childViewControllers.contains(viewController)
        
        if containsVC == false {
            self.addChildViewController(viewController)
        }
        
        setSubViewAction?(self,viewController)
        
        if containsVC == false {
            viewController.didMove(toParentViewController: self)
        }
    }
    
    func removeFromParentVC() {
        self.willMove(toParentViewController: nil)
        self.view.removeFromSuperview()
        self.removeFromParentViewController()
    }
}
